keps/sig-apps/2804-consolidate-workload-controllers-status/README.md

```diff
@@ -191,12 +225,40 @@ See the [Kubernetes API Conventions](https://git.k8s.io/community/contributors/d
+
+Since we have to deal with a workload that still could be considered healthy even when some number of replicas is down, we would like to introduce a threshold that would distinguish between these cases.
+This threshold would dictate whether the application is Functional/Operational/NominallyAvailable or not.
+The problem is that some users might consider their application to be healthy at 50%, 75%, 90% or 100% of available pods.
+
+There are two options how to implement this:
+1. Hardcode the threshold to 100% of available replicas. Then ask a community for a feedback before graduating this feature, and if there is an interest, we could introduce a field that would allow customization of this threshold.
+2. Introduce a field that would allow customizing the threshold of available replicas right from the alpha version.
+
+Default value of the threshold should be 100% to avoid false negatives.
+For example, if we would set it to 90%, the application would signal a healthy state (`status=True`) by default to the user,
+even if some replicas would be down and that would impact healthiness of the user's application.
+Also, it would be hard to change this default value later.
+
+Other possible names for the new Operational condition:
+- Functional
+- NominallyAvailable
+
+##### Implementation Details
+
+Individual workload controllers mark a ReplicaSet or StatefulSet as `Operational` when `availableReplicas >= replicas - threshold`.
+- Available replicas is a beta feature guarded by StatefulSetMinReadySeconds gate in StatefulSets, but the value defaults to ReadyReplicas when the feature gate is disabled so using it shouldn't be an issue.
+
+Implementation of `Operational` condition in Deployments with a different threshold other than 100% is not straightforward.
+There are a couple of things to consider.
+- Should it be possible to set this threshold only to a ReplicaSet, Deployment, or both ReplicaSet and Deployment? There should be a way to reconcile these values or block incompatible values on admission.
+- Should the deployment balance the threshold in a rollout across its ReplicaSets depending on the amount of replicas in each ReplicaSet? The computation of this also depends on the threshold being a number or a percentage, and if any rounding occurs.
+- Each Deployment's ReplicaSet could have a different computed status of `Operational` depending on the availability of their replicas and the threshold. Deployment thus shouldn't depend on its ReplicaSets conditions and compute its own `Operational` condition from the global state of all its managed replicas.
+
 
-DaemonSet controller marks DaemonSet as `available` when `numberUnavailable` or `desiredNumberScheduled - numberAvailable` is zero.
+DaemonSet controller marks DaemonSet as `Operational` when  `numberAvailable >= desiredNumberScheduled - threshold`.
 
 #### Batch Workloads Conditions: Waiting & Running
```

Comment:
Unfortunately, I was not involved in the previous version of the KEP, but I'm not finding much value in these conditions.

Did you consider just adding one condition named "Ready" with a value equal to `.status.ready>0`?

Reply:
I think `available` replicas is a bit more useful, since user can utilize `spec.minReadySeconds` which adds more tuning for a application to be considered healthy. They can also choose to set it to 0 to consider only the ready part.

I am not trying to tie together the batch workloads as well since their use case is quite a bit different. Do you have some special use case in mind with regards to the `Ready` condition?

Reply:
For service workloads, we want to have the highest level (captures the most actual end user impact) condition possible that allows improvements to observed outcome tuning (new fields) to show up in the value.  Ready would probably be too low level because it would not capture minReadySeconds.  I think to fully decide what we are adding we have to be very precise about the use cases the conditions solve and the end user automation problems the conditions enable.